### ISSUE TITLE
resolver hangs on certain YAML files

### ISSUE THREAD

**Issue by rohansingh** [bazelbuild/rules_k8s#290]
I have a YAML with a `ConfigMap` containing the following line:

`REDIS_URL: redis://redis:6379`

In `resolver.StringToDigest`, the `redis://redis:6379` is incorrectly interpreted as a `docker_name.Tag`. At that point, the resolver attempts to resolve the digest.

However, in our environment, `redis` resolves to a host that drops all the relevant traffic. As a result, the resolver hangs forever.

The end-user experience is pretty horrible too, since when running an `.apply` target, the resolver output is piped to `kubectl`. This makes `Ctrl-C` unusable, so you need to resort to other methods.

**Comment by rohansingh** [comment]
Possible options:

1. Add a timeout for each resolution.
2. Only try to resolve strings in a known-good list.
3. Have an option to disable this behavior entirely.

**Comment by smukherj1** [comment]
@rohansingh Thanks for filing this issue.

Looking at resolver.py myself, it seems it's original intent was to resolve docker tags. Thus, the very fact that it's attempting to resolve a redis path is an error. As for your options:

1. This just feels like a bandaid solution to me. If the link responds in a different way, such as it changes the state of the remote entity being communicated with or throws some other kind of error.
2. I like this idea. We would need to carefully filter out known docker image path formats and only attempt to resolve those.
3. I think this could be combined with 2. ie, specify a map here of keys to never resolve. Basically specify a blacklist of keys to exclude in the resolver.

@nlopezgi @chrislovecnm thoughts?

As for the kubectl thing, it would be a fundamental change in how rules_k8s works (I think). Would require some thought to design alternative solutions and come up with a design doc/implementation plan.

**Comment by rohansingh** [comment]
> This just feels like a bandaid solution to me. If the link responds in a different way, such as it changes the state of the remote entity being communicated with or throws some other kind of error.

Yup, I agree that only doing a timeout would be insufficient. I would encourage having a reasonable timeout on top of some other solution though, and just fail entirely once the timeout is hit. That would also mitigate the issue of `kubectl` hanging forever waiting for stdin.

> We would need to carefully filter out known docker image path formats and only attempt to resolve those.

I'm a bit skeptical that it could ever be possible to only attempt to resolve Docker images and never catch anything else. For example, the string `some.blackhole/foo/bar:123` is both a "valid" image name, and something I would never want resolved.

Rather, what I meant with this option was that only image names in a fixed list of strings would ever be resolved. This would also match the behavior of resolving images that the resolver publishes. Those have to be explicitly specified in the `images` dictionary, after all (otherwise we'd never know to publish them). I also think having to specify a whitelist would make the behavior less surprising.

> I think this could be combined with 2. ie, specify a map here of keys to never resolve. Basically specify a blacklist of keys to exclude in the resolver.

This is actually something I'd push strongly against. Having a blacklist like this seems fairly fragile, since any new value could re-trigger the bad behavior. If a user isn't very familiar with `rules_k8s` and is just editing the YAML for a service they own, they could potentially add something like the `some.blackhole` string I mentioned above, and then be stuck.

Also, it'd be hard to know what to blacklist. While with a whitelist I could specify images that I know should resolve, it's hard to know which "bad" strings might actually resolve to real locations in a different user's environment. In practice, I'd have to blacklist every string that is not a known-good image.

Really, just being able to disable the whole resolution of images that haven't been built/published by Bazel would be the best. That actually dovetails well with a whitelist, since "disable" would just be providing an empty whitelist.

> As for the kubectl thing, it would be a fundamental change in how rules_k8s works (I think). Would require some thought to design alternative solutions and come up with a design doc/implementation plan.

I'm guessing what's happening with `Ctrl-C` is that the resolver is shutting down on `SIGINT`, but `kubectl` isn't properly handling `SIGINT` while waiting for stdin. If that is the case (would need to be verified), couldn't we just write the output of the resolver to a temp file and then feed that to `kubectl -f <file>`?

That way, if the resolver ever hangs, `Ctrl-C` will kill it and the whole surrounding Bash script.

**Comment by rohansingh** [comment]
One thing I don't understand is why the resolver tries to walk and resolve *all* YAML strings as potential Docker images. Are there use cases outside of a dictionary item where the key is "image" (like in a `Container` definition)?

**Comment by Globegitter** [comment]
Seems #185 and #252 might be somewhat related to this.

**Comment by rohansingh** [comment]
While #252 is related to the `Ctrl-C` problem I mentioned, it's also unrelated enough that it should be fixable before this issue is resolved. Does anyone see any harm in doing what I suggested above and writing resolver output to temp file instead of piping directly to `kubectl`?

**Comment by nlopezgi** [comment]
sorry for the delay getting back. This is a gnarly issue and we've had some discussions about what to do but have not yet come to a conclusion. For now, I think @rohansingh suggestion of  write the output of the resolver to a temp file should help. I'd be happy to review a PR with that change. 
As for the wider question. My opinion is that we should allow disabling resolution of tags altogether if the user requests it (i.e., not whitelist, no blacklist, its all or nothing). What do other folks think about that suggestion?

**Comment by rohansingh** [comment]
Internally we've forked the resolver and just removed the tag resolution functionality. So I'd be happy with just having a "disable" option. Though also we fork the resolver anyway for other reasons, so I don't really have a horse in this race.

**Comment by Globegitter** [comment]
Just out of curiosity, what are the other reasons for forking?

**Comment by rohansingh** [comment]
@Globegitter Really just to implement #223.
